### PR TITLE
[Sema] Fix name confusion in Decodable synthesis

### PR DESCRIPTION
**Pull request by beccadax**
`Decodable`’s `init(from:)` synthesis sometimes mistook a static property for an identically-named instance property, which could cause it to skip a property or possibly make other mistakes. This change factors a common helper function from `encode(to:)` and `init(from:)` synthesis which implements the right logic for both.

Resolves [SR-10045](https://bugs.swift.org/browse/SR-10045) a.k.a. rdar://problem/48702779.

### CODE REVIEW

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 1]
```diff
@@ -509,6 +509,49 @@ static CallExpr *createContainerKeyedByCall(ASTContext &C, DeclContext *DC,
...
+  ASTContext &C = elt->getASTContext();
+
+  for (auto decl : targetDecl->lookupDirect(DeclName(elt->getName()))) {
+    if (auto *vd = dyn_cast<VarDecl>(decl)) {
+      if (!vd->isStatic()) {
+        // This is the VarDecl we're looking for.
+
+        auto varType = conformanceDC->mapTypeIntoContext(vd->getInterfaceType());
+
+        if (auto referenceType = varType->getAs<ReferenceStorageType>()) {
```
Comment (theblixguy) [inline lib/Sema/DerivedConformanceCodable.cpp]: You can do `varType-> getValueInterfaceType()` and get rid of this if check.

Comment (brentdax) [reply lib/Sema/DerivedConformanceCodable.cpp]: Thanks for the tip! I'm testing that change now.

**Comment by brentdax** [comment]
@swift-ci please smoke test

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 2]
```diff
@@ -831,42 +863,22 @@ static void deriveBodyDecodable_init(AbstractFunctionDecl *initDecl, void *) {
...
         continue;
 
-      // Potentially unwrap a layer of optionality from the var type. If the var
-      // is Optional<T>, we want to decodeIfPresent(T.self, forKey: ...);
-      // otherwise, we can just decode(T.self, forKey: ...).
-      // This is also true if the type is an ImplicitlyUnwrappedOptional.
-      auto varType = conformanceDC->mapTypeIntoContext(
-          varDecl->getInterfaceType());
-      auto methodName = C.Id_decode;
-      if (auto referenceType = varType->getAs<ReferenceStorageType>()) {
```
Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: I don't see this unwrapping for storage types in the above function — is it no longer necessary, or are we missing unit tests which would have caught this?

File: lib/Sema/DerivedConformanceCodable.cpp [hunk 3]
```diff
@@ -509,6 +509,50 @@ static CallExpr *createContainerKeyedByCall(ASTContext &C, DeclContext *DC,
...
+              dyn_cast<BoundGenericType>(varType->getCanonicalType());
+          varType = boundOptionalType->getGenericArgs()[0];
+        }
+
+        return std::make_tuple(vd, varType, useIfPresentVariant);
+      }
+    }
+  }
+
+  llvm_unreachable("Should have found at least 1 var decl");
```
Comment (itaiferber) [inline lib/Sema/DerivedConformanceCodable.cpp]: I'm assuming this will assert if ever hit?

**Review by itaiferber — COMMENTED** [review]
Changes look reasonable overall, just two small questions:

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 2 above)
Comment (brentdax) [reply lib/Sema/DerivedConformanceCodable.cpp]: The `mapTypeIntoContext()` line changed slightly when I moved it:

```swift
conformanceDC->mapTypeIntoContext(vd->getValueInterfaceType());
```

`getValueInterfaceType()` (as opposed to `getInterfaceType()` incorporates the `ReferenceStorageType` unwrapping.

File: lib/Sema/DerivedConformanceCodable.cpp (hunk 3 above)
Comment (brentdax) [reply lib/Sema/DerivedConformanceCodable.cpp]: Yes.

**Comment by brentdax** [comment]
@swift-ci please smoke test and merge

**Comment by brentdax** [comment]
@swift-ci please smoke test and merge

**Pull request merged**